### ISSUE TITLE
Empty "state" Value in AppHomeOpened Event Causes Parsing Error

### ISSUE THREAD

**Issue by HunterL** [slack-go/slack#943]
### What happened
- Published a home view with some text and a simple static_select element
- Every subsequent "AppHomeOpened" event has been parsed as "err_bad_message"

` &{Cause:parsing Events API event: EventsAPI Error parsing inner event: unmarshalling_error, 
json: cannot unmarshal array into Go struct field ViewState.view.state.values of type map[string]map[string]slack.BlockAction
`

The client is recieving the following JSON (as part of a larger event...):
```
"state": {
    "values": []
},
```
And attempting to marshal it into

```
type ViewState struct {
	Values map[string]map[string]BlockAction `json:"values"`
}
```

### Versions
- Go: 1.15.5
- slack-go/slack: 0.9.1

### Work Around
Thought I should add an edit on how we're working around this until there is clarity from Slack and we can merge a fix into slack-go
1. Add a struct to help us pull out the payload and envelope ID
```
// HomeEvent is a helper struct for recovery from err bad message
type HomeEvent struct {
	Envelope string                     `json:"envelope_id"`
	Payload  json.RawMessage `json:"payload"`
}
```
2. Pull out the payload, fix the brackets causing an issue, stuff it back into an event, push back into the client channel
```
case socketmode.EventTypeErrorBadMessage:
	/*
		We do some handling here to attempt to recover the event from a known issue with "AppHomeOpened"
		If we fail to recover the assumption is that the bad event was for a legitimate reason and not
		an internal issue with slack-go
	*/
	var e *socketmode.ErrorBadMessage
	var ok bool
	if e, ok = event.Data.(*socketmode.ErrorBadMessage); !ok {
		log.Printf("Bad Message Not Cast: %+v", event)
		continue
	}
	var rawBytes []byte
	if rawBytes, err = e.Message.MarshalJSON(); err != nil {
		log.Printf("Bad Message Not Marshalled. Err: %+v\n Event: %+v", err, event)
		continue
	}

	/*
		This line replaces `"state":{"values":[]}` with `"state":{"values":{}}`
		The latter is parsed correctly by slack-go while the former is what kicks off the "ErrorBadMessage"
		event in the first place. It is, unfortunately, the best way to fix this behavior until there is
		clarification from slack on intended return values and a fix is merged into slack-go
	*/
	rawMessage := bytes.Replace(rawBytes, []byte{34, 115, 116, 97, 116, 101, 34, 58, 123, 34, 118, 97, 108, 117, 101, 115, 34, 58, 91, 93, 125}, []byte{34, 115, 116, 97, 116, 101, 34, 58, 123, 34, 118, 97, 108, 117, 101, 115, 34, 58, 123, 125, 125}, 1)
	var hE HomeEvent
	if err := json.Unmarshal(rawMessage, &hE); err != nil {
		log.Printf("Raw Message Not Marshalled: %s", err)
                 continue
	}

	// Parse the raw json payload without verifying the token as it will fail otherwise
	var newEvent slackevents.EventsAPIEvent
	if newEvent, err = slackevents.ParseEvent(hE.Payload, slackevents.OptionNoVerifyToken()); err != nil {
		log.Printf("Bad Message Not Parsed. Err: %+v\n Inner JSON: %+v", err, rawMessage)
		continue
	}

	// Plug all of our parts into an event
	fabEvent := socketmode.Event{
		Type: socketmode.EventTypeEventsAPI,
		Data: newEvent,
		Request: &socketmode.Request{ //  we need to attach the envelope ID for the request to Ack
			Type:       "events_api",
			EnvelopeID: hE.Envelope,
		},
	}
	h.Client.Events <- fabEvent
```
edit: Realized there is some ambiguity on "h.Client.Events". The "h.Client" is of type `*socketmode.Client`

**Comment by HunterL** [comment]
Dug into this some more, the issue seems that slack-go expects the empty state be a nil JSON obj and not an empty JSON array.
```
"state": {
    "values": {} // <-- What we expect  || What we get --> // "values":[]
},
```

You can test this by modifying `slackevents/parsers.go` and replacing the bytes of the raw JSON before it gets marshaled
```
func parseInnerEvent(e *EventsAPICallbackEvent) (EventsAPIEvent, error) {
	iE := &slack.Event{}
	rawBytes, _ := e.InnerEvent.MarshalJSON()
	rawInnerJSON := bytes.Replace(rawBytes, []byte{34, 115, 116, 97, 116, 101, 34, 58, 123, 34, 118, 97, 108, 117, 101, 115, 34, 58, 91, 93, 125}, []byte{34, 115, 116, 97, 116, 101, 34, 58, 123, 34, 118, 97, 108, 117, 101, 115, 34, 58, 123, 125, 125}, 1)
...
```

**Comment by HunterL** [comment]
Sent an email to slack support on this one because between what I'm receiving in responses, what they document, and what their API tester returns - the expected return behavior (nil json obj or empty array) is unclear. Will update once cleaned up.

**Comment by HunterL** [comment]
Heard back from Slack support, it is a bug.

```
I'm afraid we have a bug with apps that have Socket Mode enabled where the state value in the app_home_opened event payload returns a values array instead of an object.

Please note that it may take a bit of time for the team to track down the source of this problem which is needed before we can look into how to get this issue resolved.

My apologies for the trouble, and that I don't have better news for you on this at the moment. We appreciate your patience on this. Please don't hesitate to reach out if there's anything else that I can help with in the meantime.
```

**Comment by kanata2** [comment]
@HunterL Thank you for your investigation.

**Comment by HunterL** [comment]
Received a note from Slack support that this issue has been fixed.

**Comment by derekritchison** [comment]
Hey there- I am actually seeing this _exact_ behavior, and it started happening out of nowhere mid-testing earlier this morning. Up until recently I had been receiving full view_submission responses with view.state.values containing my submitted text responses. But suddenly I started receiving:

"state": {
    "values": []
},

I am incredibly confused because I did not change anything in my original views.open modal. The response JSON... just changed?

Happy to share any additional info to help troubleshoot what could have happened here.

**Comment by kanata2** [comment]
I re-open this issue for investigation. Thanks.

**Comment by nlopes** [comment]
This has ping ponged a bit but I'm going to close this under the assumption that Slack has sorted this. I couldn't replicate but I totally get this has happened and when it does is a breaking change.

My view is that I'd rather try to keep this library as simple as possible and not super defensive towards slack APIs reverts.
If anyone still sees this issue in their environment, please let me know and I'll go in the other direction.
